x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ConfigurableClusterPrivileges.java

```diff
@@ -362,4 +398,276 @@ private interface Fields {
+                    + Arrays.toString(indexPatterns())
+                    + ":"
+                    + Fields.PRIVILEGES
+                    + ":"
+                    + Arrays.toString(privileges())
+                    + "}";
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                ManageRolesIndexPermissionGroup that = (ManageRolesIndexPermissionGroup) o;
+                return Arrays.equals(indexPatterns, that.indexPatterns) && Arrays.equals(privileges, that.privileges);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(Arrays.hashCode(indexPatterns), Arrays.hashCode(privileges));
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "{"
+                + getCategory()
+                + ":"
+                + Fields.MANAGE.getPreferredName()
+                + ":"
+                + Fields.INDICES.getPreferredName()
+                + "=["
+                + Strings.collectionToDelimitedString(indexPermissionGroups, ",")
+                + "]}";
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final ManageRolesPrivilege that = (ManageRolesPrivilege) o;
+
+            if (this.indexPermissionGroups.size() != that.indexPermissionGroups.size()) {
+                return false;
+            }
+
+            for (int i = 0; i < this.indexPermissionGroups.size(); i++) {
+                if (Objects.equals(this.indexPermissionGroups.get(i), that.indexPermissionGroups.get(i)) == false) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(indexPermissionGroups.hashCode());
+        }
+
+        @Override
+        public ClusterPermission.Builder buildPermission(final ClusterPermission.Builder builder) {
+            return builder.addWithPredicateSupplier(
+                this,
+                Set.of(PutRoleAction.NAME, ActionTypes.BULK_PUT_ROLES.name(), ActionTypes.BULK_DELETE_ROLES.name(), DeleteRoleAction.NAME),
+                requestPredicateSupplier
+            );
+        }
+
+        private static boolean requestIndexPatternsAllowed(
+            IndicesPermission indicesPermission,
+            String[] requestIndexPatterns,
+            String[] privileges
+        ) {
+            return indicesPermission.checkResourcePrivileges(Set.of(requestIndexPatterns), false, Set.of(privileges), true, null);
+        }
+
+        private static boolean hasNonIndexPrivileges(RoleDescriptor roleDescriptor) {
```

Comment:
Is it intentional that this method is not checking `roleDescriptor.hasConfigurableClusterPrivileges()` as well?
Meaning should a user with global manage role be able to create, update or delete roles with global privileges?

Reason why I'm asking is because I think allowing this could potentially lead to a privilege escalation. The flow I have in mind is that a user with the global manage roles privilege could create a role which includes global privileges that are more powerful than what user have. For example:

If a user has following global manage roles:

```json
POST _security/role/manage-roles
{
    "global": {
        "role": {
            "manage": {
                "indices": [
                    {
                        "names": ["allowed-index-prefix-*"],
                        "privileges":["read"]
                    }
                ]
            }
        }
    }
}
```
it could potentially create below role that grants global privileges across all indices/aliases:

```json
POST _security/role/my-new-manage-role
{
    "global": {
        "role": {
            "manage": {
                "indices": [
                    {
                        "names": ["*"],
                        "privileges":["all"]
                    }
                ]
            }
        }
    },
    "indices": [
        {
            "names": ["allowed-index-prefix-*"],
            "privileges":["read"]
        }
    ]
}
```

The `my-new-manage-role` can then be used to manage and delete all other roles. This is possible because we only check `indices` part and do not check `global.manage.indices`.

Reply:
Good catch! No, that's not intentional. I thought the `hasClusterPrivileges` covered that too, but it doesn't. I'll update the code.

Reply:
Added code + test to handle it